File: src/monster_bootstrap.pl
:- module(monster_bootstrap,
    [ new_monster_context/3,
      monster_manager_execute/3,
      calendar_date_spaghetti/3
    ]).

new_monster_context(CalculationDay, TargetDay,
    monster_context(
        CalculationDay,
        TargetDay,
        bootstrap,
        entry,
        authoritative,
        new,
        0,
        none,
        none,
        [],
        metrics(0,0),
        [],
        none
    )).

monster_manager_execute(CalculationDay, TargetDay, FinalContext) :-
    new_monster_context(CalculationDay, TargetDay, Context0),
    base_dispatch(entry, Context0, FinalContext).

base_dispatch(entry, Context0, FinalContext) :-
    context_with_phase(Context0, validation, entered, Context1),
    base_dispatch(validation, Context1, FinalContext).
base_dispatch(validation, Context0, FinalContext) :-
    Context0 = monster_context(CalculationDay,TargetDay,_,_,_,_,_,_,_,Trace,Metrics,Logs,Error),
    ( integer(CalculationDay), integer(TargetDay) ->
        Context1 = monster_context(CalculationDay,TargetDay,bootstrap,ready,authoritative,
                                   validated,0,base_validator,none,
                                   [validated|Trace],Metrics,Logs,Error),
        base_dispatch(ready, Context1, FinalContext)
    ; Context1 = monster_context(CalculationDay,TargetDay,bootstrap,failed,authoritative,
                                 validation_error,0,base_validator,none,
                                 [validation_failed|Trace],Metrics,Logs,invalid_day_input),
      base_dispatch(failed, Context1, FinalContext)
    ).
base_dispatch(ready, Context0, FinalContext) :-
    bump_dispatch_metric(Context0, Context1),
    context_with_status(Context1, bootstrap_ready, FinalContext).
base_dispatch(failed, Context, Context).

context_with_phase(
    monster_context(C,T,_,_,M,_,R,H,P,Trace,Metrics,Logs,Error),
    Phase, Status,
    monster_context(C,T,bootstrap,Phase,M,Status,R,H,P,[Phase|Trace],Metrics,Logs,Error)).

context_with_status(
    monster_context(C,T,Boot,Phase,Mode,_,R,H,P,Trace,Metrics,Logs,Error),
    Status,
    monster_context(C,T,Boot,Phase,Mode,Status,R,H,P,Trace,Metrics,Logs,Error)).

bump_dispatch_metric(
    monster_context(C,T,Boot,Phase,Mode,Status,R,H,P,Trace,metrics(Dispatches,Failures),Logs,Error),
    monster_context(C,T,Boot,Phase,Mode,Status,R,H,P,Trace,metrics(Dispatches2,Failures),Logs,Error)) :-
    Dispatches2 is Dispatches + 1.

calendar_date_spaghetti(CalculationDay, TargetDay, _) :-
    monster_manager_execute(CalculationDay, TargetDay, Context),
    Context = monster_context(_,_,_,_,_,bootstrap_ready,_,_,_,_,_,_,_),
    throw(error(stage_not_available(54), calendar_date_spaghetti/3)).

File: src/monster_bootstrap.m
:- module monster_bootstrap.
:- interface.

:- import_module integer.
:- import_module list.

:- type z == integer.

:- type phase
    --->    bootstrap_entry
    ;       bootstrap_validate
    ;       bootstrap_ready.

:- type monster_context
    --->    monster_context(
                calculation_day :: z,
                target_day      :: z,
                phase           :: phase,
                status_code     :: string,
                branch_trace    :: list(string),
                metric_events   :: list(string),
                log_events      :: list(string),
                validation_codes:: list(string)
            ).

:- type dispatch_result
    --->    dispatch_ok(monster_context)
    ;       dispatch_error(string, monster_context).

:- func new_context(z, z) = monster_context.
:- func base_dispatch(monster_context) = dispatch_result.
:- pred validate_bootstrap_context(monster_context::in) is semidet.

:- implementation.

new_context(C, T) = monster_context(
    C,
    T,
    bootstrap_entry,
    "BOOTSTRAP_NEW",
    ["BOOTSTRAP_ENTER"],
    [],
    [],
    []
).

base_dispatch(Context0) = Result :-
    Context1 = Context0 ^ phase := bootstrap_validate,
    Context2 = Context1 ^ branch_trace :=
        Context1 ^ branch_trace ++ ["BOOTSTRAP_VALIDATE"],
    ( if validate_bootstrap_context(Context2) then
        Context3 = Context2 ^ phase := bootstrap_ready,
        Context4 = Context3 ^ status_code := "BOOTSTRAP_READY",
        Context5 = Context4 ^ metric_events :=
            Context4 ^ metric_events ++ ["bootstrap.ready"],
        Result = dispatch_ok(Context5)
    else
        Context3 = Context2 ^ status_code := "BOOTSTRAP_INVALID",
        Context4 = Context3 ^ validation_codes :=
            Context3 ^ validation_codes ++ ["BOOTSTRAP_CONTEXT_INVALID"],
        Result = dispatch_error("BOOTSTRAP_CONTEXT_INVALID", Context4)
    ).

validate_bootstrap_context(Context) :-
    Context ^ status_code = Status,
    Status \= "".
